### PR TITLE
Migrate Jupyter Bindings of each module to this repo, so they can be maintained along with the modules.

### PR DESCRIPTION
**Pull request by kenns29**
Migrate the Jupyter Bindings (https://github.com/uber/manifold/issues/26)

### CODE REVIEW

File: bindings/jupyter-modules/lerna.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "lerna": "3.14.1",
+  "version": "0.5.2",
```
Comment (gnavvy) [inline bindings/jupyter-modules/lerna.json]: Is it possible to share lerna.json to align the versioning between the component and its binding?

File: bindings/jupyter-modules/modules/jupyter-graph-builder/package.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "@mlvis/jupyter-graph-builder",
+  "version": "0.0.1",
```
Comment (gnavvy) [inline bindings/jupyter-modules/modules/jupyter-graph-builder/package.json]: nit: should start with 1.0.0 but will be synced once we use lerna to bump the versions.

File: bindings/jupyter-modules/modules/jupyter-manifold/package.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  "version": "0.0.1",
+  "main": "dist/index.js",
+  "scripts": {
+    "build": "webpack --progress",
+    "watch": "webpack --watch --progress"
+  },
+  "dependencies": {
+    "global": "^4.4.0",
+    "react": "^16.8.0",
+    "react-dom": "^16.9.0",
```
Comment (gnavvy) [inline bindings/jupyter-modules/modules/jupyter-manifold/package.json]: align react versions (to 16.9 if possible, better hooks support)

File: bindings/jupyter-modules/package.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "jupyter-packages",
+  "description": "A set of jupyter wrappers",
+  "version": "0.0.1",
+  "private": true,
```
Comment (gnavvy) [inline bindings/jupyter-modules/package.json]: remove this after moving over?

**Comment by kenns29** [comment]
Use root level lerna to build jupyter-modules

File: bindings/jupyter-modules/modules/jupyter-manifold/package.json (hunk 3 above)
Comment (kenns29) [reply bindings/jupyter-modules/modules/jupyter-manifold/package.json]: bumped

**Comment by kenns29** [comment]
Use Apache License 2.0 for both Python and Javascript packages.

File: bindings/jupyter/docs/manifold.html [hunk 5]
```diff
@@ -0,0 +1,13400 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (Firenze11) [reply bindings/jupyter/docs/manifold.html]: How are these files generated?

[the remark this answers is not in the document]
Comment (kenns29) [reply bindings/jupyter/docs/manifold.html]: use bindings/jupyter/docs/make-docs.sh:
```
#!/bin/bash
jupyter nbconvert --to html ../notebooks/*.ipynb
jupyter nbconvert --to markdown ../notebooks/*.ipynb
mv ../notebooks/*.html .
mv ../notebooks/*.md .
```

File: bindings/jupyter-modules/jupyter-multi-way-plot/package.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  },
+  "main": "dist/index.js",
+  "scripts": {
+    "watch": "webpack --watch --progress",
+    "clean": "rm -rf dist",
+    "copy": "cp -r package.json dist",
+    "build": "yarn clean && yarn transpile",
+    "transpile": "NODE_ENV=production babel src -d dist -s --config-file ../../../babel.config.js && yarn copy"
+  },
+  "dependencies": {
```
Comment (Firenze11) [inline bindings/jupyter-modules/jupyter-multi-way-plot/package.json]: has to depend on multi-way-plot

Comment (kenns29) [reply bindings/jupyter-modules/jupyter-multi-way-plot/package.json]: Done.

**Comment by chrisirhc** [comment]
Could you please update the PR title to something more descriptive of the change?
The PR's description should follow git commit log guidelines. See https://chris.beams.io/posts/git-commit/ for an example

File: bindings/jupyter/setup.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+INSTALL_REQUIRES = [
+    'traitlets',
+    'ipywidgets',
+    'pandas'
+]
+
+
+DEFAULT_JREQUIREMENTS = {
+    "FeatureListView": {
+        "@uber/feature-list-view": "external"
```
Comment (Firenze11) [inline bindings/jupyter/setup.py]: change to `@mlvis`

File: modules/graph-builder/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# GraphBuilder
+
+A generic node-link diagram
+
+## Usage
+
+Example:
+```js
+import GraphBuilder from '@uber/graph-builder';
```
Comment (Firenze11) [inline modules/graph-builder/README.md]: change to `@mlvis`

File: modules/stacked-calendar/README.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+The _Stacked Calendar_ component aggregates hourly-stamped values and visualizes the results in the form of a calendar.
+
+The columns and rows are corresponding to the day_of_week and hour_of_day, respectively.\
+The data values of each cell (hour) are aggregated, percentiled, and mapped to color gradients to reveal the value distributions.
+
+## Usage
+
+```js
+import StackedCalendar from '@uber/stacked-calendar';
```
Comment (Firenze11) [inline modules/stacked-calendar/README.md]: change to `@mlvis`

File: modules/stacked-calendar/README.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+|              167               |  0.4  |
+
+## Installation
+
+To install the dependencies from NPM:
+
+```bash
+npm install @uber/stacked-calendar
+# or
+yarn install @uber/stacked-calendar
```
Comment (Firenze11) [inline modules/stacked-calendar/README.md]: change to `@mlvis`

**Comment by Firenze11** [comment]
> Updated jupyter-modules structure:
> 
> ```
> jupyter-modules/
> ├── jupyter-graph-builder
> │   ├── dist
> │   ├── package.json
> │   └── src
> ├── jupyter-manifold
> │   ├── dist
> │   ├── package.json
> │   └── src
> └── jupyter-multi-way-plot
>     ├── dist
>     ├── package.json
>     └── src
> ```
> 
> Each module now has the following build script to allow them to be built by the root lerna because ocular cannot build modules in deeper level subfolders.
> 
> ```json
>  "scripts": {
>     "watch": "webpack --watch --progress",
>     "clean": "rm -rf dist",
>     "copy": "cp -r package.json dist",
>     "build": "yarn clean && yarn transpile",
>     "transpile": "NODE_ENV=production babel src -d dist -s --config-file ../../../babel.config.js && yarn copy"
> },
> ```
You mean root lerna in jupyter-modules/ ? Or in the repo root? 
If latter, which script allows repo root lerna to access bindings/jupyter-mocules/ ? Also, shall we make separate scripts in root to differentiate building modules/ and jupyter-modules/ ?

File: bindings/jupyter/docs/manifold.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (Firenze11) [reply bindings/jupyter/docs/manifold.html]: Got it. I think considering the size of this html file, I'm not sure whether we need to commit it right now. I suggest to commit it when we discover it's definitely needed after we have a doc website. Also, we could choose to commit it in a different branch (like what we did with transpiled javascript in `gh-pages` branch

**Comment by kenns29** [comment]
> > Updated jupyter-modules structure:
> > ```
> > jupyter-modules/
> > ├── jupyter-graph-builder
> > │   ├── dist
> > │   ├── package.json
> > │   └── src
> > ├── jupyter-manifold
> > │   ├── dist
> > │   ├── package.json
> > │   └── src
> > └── jupyter-multi-way-plot
> >     ├── dist
> >     ├── package.json
> >     └── src
> > ```
> > 
> > 
> > Each module now has the following build script to allow them to be built by the root lerna because ocular cannot build modules in deeper level subfolders.
> > ```json
> >  "scripts": {
> >     "watch": "webpack --watch --progress",
> >     "clean": "rm -rf dist",
> >     "copy": "cp -r package.json dist",
> >     "build": "yarn clean && yarn transpile",
> >     "transpile": "NODE_ENV=production babel src -d dist -s --config-file ../../../babel.config.js && yarn copy"
> > },
> > ```
> 
> You mean root lerna in jupyter-modules/ ? Or in the repo root?
> If latter, which script allows repo root lerna to access bindings/jupyter-mocules/ ? Also, shall we make separate scripts in root to differentiate building modules/ and jupyter-modules/ ?

pakcage.json workspace settings along with lerna.json settings such as this:
```
{
  "lerna": "3.14.1",
  "version": "1.0.0",
  "packages": ["modules/*", "bindings/jupyter-modules/*"],
  "npmClient": "yarn",
  "useWorkspaces": true
}
```
enables using root lerna to build both modules and jupyter-modules.

**Comment by Firenze11** [comment]
> > > Updated jupyter-modules structure:
> > > ```
> > > jupyter-modules/
> > > ├── jupyter-graph-builder
> > > │   ├── dist
> > > │   ├── package.json
> > > │   └── src
> > > ├── jupyter-manifold
> > > │   ├── dist
> > > │   ├── package.json
> > > │   └── src
> > > └── jupyter-multi-way-plot
> > >     ├── dist
> > >     ├── package.json
> > >     └── src
> > > ```
> > > 
> > > 
> > > Each module now has the following build script to allow them to be built by the root lerna because ocular cannot build modules in deeper level subfolders.
> > > ```json
> > >  "scripts": {
> > >     "watch": "webpack --watch --progress",
> > >     "clean": "rm -rf dist",
> > >     "copy": "cp -r package.json dist",
> > >     "build": "yarn clean && yarn transpile",
> > >     "transpile": "NODE_ENV=production babel src -d dist -s --config-file ../../../babel.config.js && yarn copy"
> > > },
> > > ```
> > 
> > 
> > You mean root lerna in jupyter-modules/ ? Or in the repo root?
> > If latter, which script allows repo root lerna to access bindings/jupyter-mocules/ ? Also, shall we make separate scripts in root to differentiate building modules/ and jupyter-modules/ ?
> 
> pakcage.json workspace settings along with lerna.json settings such as this:
> 
> ```
> {
>   "lerna": "3.14.1",
>   "version": "1.0.0",
>   "packages": ["modules/*", "bindings/jupyter-modules/*"],
>   "npmClient": "yarn",
>   "useWorkspaces": true
> }
> ```
> 
> enables using root lerna to build both modules and jupyter-modules.
I think it's better to differentiate the 2 levels of build. Something like (in root package.json):
```
"build": "ocular-build",
"build:all": "ocular-build && cd bindings/jupyter-modules && yarn build && cd ../"
```
Then in `bindings/jupyter-modules/package.json` you implement the second level lerna.

**Comment by kenns29** [comment]
> "build:all": "ocular-build && cd bindings/jupyter-modules && yarn build && cd ../"

This won't work. because 
1) The lerna command has to execute at root to leverage root level lerna.json
2) if for lerna to identify dependencies within the monorepo (e.g. `@mlvis/manifold`), they have to be built together.

File: bindings/jupyter/docs/manifold.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (kenns29) [reply bindings/jupyter/docs/manifold.html]: removed the htmls

File: bindings/jupyter-modules/jupyter-graph-builder/package.json [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  "main": "dist/index.js",
+  "scripts": {
+    "watch": "webpack --watch --progress",
+    "clean": "rm -rf dist",
+    "copy": "cp -r package.json dist",
+    "build": "yarn clean && yarn transpile",
+    "transpile": "NODE_ENV=production babel src -d dist -s --config-file ../../../babel.config.js && yarn copy"
+  },
+  "dependencies": {
+    "react": "^16.8.0"
```
[the remark this answers is not in the document]
Comment (kenns29) [reply bindings/jupyter-modules/jupyter-graph-builder/package.json]: Done.

File: bindings/jupyter/js/package.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    "build": "webpack",
+    "watch": "node --max_old_space_size=8192 node_modules/.bin/webpack --progress --watch"
+  },
+  "dependencies": {
+    "global": "^4.4.0",
+    "react": "^16.8.0",
+    "reselect": "^4.0.0"
+  },
+  "devDependencies": {
+    "@babel/cli": "^7.1.2",
```
[the remark this answers is not in the document]
Comment (kenns29) [reply bindings/jupyter/js/package.json]: This is a challenge with haven't yet figure out how to address. Lerna can pick up deps in the repo if they are in the same workspace. But for those that outside of the workspace, e.g. website, example, jupyter js, etc. There may need to be some additional settings in order for the modules to be picked up.

File: bindings/jupyter/jrequirements.json [hunk 13]
```diff
@@ -0,0 +1,17 @@
+{
+  "FeatureListView": {
+    "@uber/feature-list-view": "external"
```
[the remark this answers is not in the document]
Comment (kenns29) [reply bindings/jupyter/jrequirements.json]: Done.

File: modules/graph-builder/README.md (hunk 8 above)
Comment (kenns29) [reply modules/graph-builder/README.md]: Done.

File: modules/stacked-calendar/README.md (hunk 10 above)
Comment (kenns29) [reply modules/stacked-calendar/README.md]: Done.

File: bindings/jupyter/setup.py (hunk 7 above)
Comment (kenns29) [reply bindings/jupyter/setup.py]: Done.

File: modules/stacked-calendar/README.md (hunk 9 above)
Comment (kenns29) [reply modules/stacked-calendar/README.md]: Done

**Pull request merged**